From 0794c9c54702f3def8998e7153cf12c5d4a2d482 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:42:28 -0400
Subject: [ethernet] cxgb4: LE-Workaround is not atomic in firmware

Message-id: <1410475447-94392-10-git-send-email-svemuri@redhat.com>
Patchwork-id: 93871
O-Subject: [RHEL7.1 PATCH BZ 1124947 009/108] cxgb4: LE-Workaround is not atomic in firmware
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

The LE workaround in firmware is not atomic and fw_ofld_connection_wrs must not interleave.
Therefore, when the workaround is enabled, we need to send all ctrlq WRs on a single ctrl queue.

Based on original work by Santosh Rastapur <santosh@chelsio.com>

Signed-off-by: Kumar Sanghvi <kumaras@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 4fe44dd77602fef804561c46ad10b988ee2859ac)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/chelsio/cxgb4/sge.c b/drivers/net/ethernet/chelsio/cxgb4/sge.c
index cc3511a..9b9964e 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/sge.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/sge.c
@@ -1467,8 +1467,12 @@ static inline int ofld_send(struct adapter *adap, struct sk_buff *skb)
 {
  unsigned int idx = skb_txq(skb);
 
- if (unlikely(is_ctrl_pkt(skb)))
+ if (unlikely(is_ctrl_pkt(skb))) {
+  /* Single ctrl queue is a requirement for LE workaround path */
+  if (adap->tids.nsftids)
+   idx = 0;
   return ctrl_xmit(&adap->sge.ctrlq[idx], skb);
+ }
  return ofld_xmit(&adap->sge.ofldtxq[idx], skb);
 }
 
-- 
1.7.1